Btdag

webgisgeek opened the pull request:
## What this pull request accomplishes:

- review and merge changes

## Issue(s) this solves:

- # documentation on creating new bt routes

## What, in particular, needs to reviewed:

-

Commit:
#343 testing first DAG

Commit:
#343 test run with new date

Commit:
#343 test run reader_history table

Commit:
343 test run reader history table again

Commit:
#250 test dag for broken readers

Commit:
#250 test dag corrected

Commit:
250 test dag updated again

Commit:
250 test dag updated test

Commit:
#250 test dag updated again

Commit:
#317 readme file updated

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,81 @@
+logging.basicConfig(level=logging.DEBUG)
+
+#To connect to pgadmin
+CONFIG = configparser.ConfigParser()
```
I should set up a bt bot account for Airflow so then the credentials would be imported 
https://github.com/CityofToronto/bdit_data-sources/blob/master/dags/pull_wys.py#L59-L60

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,81 @@
+from psycopg2.extras import execute_values
+from psycopg2 import connect, Error
+import logging
+import configparser
+
+LOGGER = logging.getLogger(__name__)
+logging.basicConfig(level=logging.DEBUG)
```
Airflow is already taking care of the `logging` configuration, the code should simply call `logging.info`, etc, further down.

https://stackoverflow.com/a/41841689/4047679

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,81 @@
+def task_fail_slack_alert(context):
+    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
+    slack_msg = """
+            :red_circle: A bt Reader may be broken!
+            *Task*: {task}  
+            *Dag*: {dag} 
+            *Execution Time*: {exec_date}  
+            *Log Url*: {log_url} 
```
This should be updated to our more recent Slack message style.

See for example

https://github.com/CityofToronto/bdit_data-sources/blob/master/dags/pull_wys.py#L28-L44

radumas commented on dags/bt_read_history.py:
```diff
@@ -0,0 +1,60 @@
+
+## Postgres update bluetooth reader_status_history DAG
+# Task to update the reader_status_history table daily
+bt_status_history = PostgresOperator(sql='''SELECT * from mohan.reader_status_history(current_date)''',
```
This isn't going to work because `current_date` isn't a template parameter.

See for example https://github.com/CityofToronto/bdit_data-sources/blob/master/dags/refresh_wys_monthly.py#L72 and our documentation "**How to use {{ ds }}**"

In short it should just be `{{ ds }}` here

radumas reviewed (changes requested):
Some suggestions on the DAGs, also wondering where the `sql` is.

Commit:
#343 blip pipeline fail check all in one DAG

Commit:
#343 blip pipeline fail check all in one DAG start date updated

radumas commented:
@KatiRG @chmnata if you could review the route updating documentation here that would be great.

webgisgeek commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,81 @@
+
+def bad_readers(con, run_date):
+    with con.cursor() as cursor: 
+        select_query = """SELECT reader from mohan.broken_reader()"""
```
forgot about SQL altogether. Will add it to the branch too. Yes date should be a parameter. Will change that

Commit:
#381 SQL add in create_tables folder

Commit:
#343 SQL for functions used in airflow  added in functions folder

KatiRG commented:
@webgisgeek Hi Mohan, the readme is pretty clear, I edited a few minor things but there were some things that I didn't understand:

- _In this update, there exists an Excel Sheet template that contains details of newly added bluetooth detectors (`update/img/template.PNG`). The details include proposed route name, description, intersection name (BDIT convention), sensor id and lat/lon at start point, and sensor id and lat/lon at the end points along with numerous other fields._ 
=> there is no `sensor id` column in the `Start point` section

- _The routes that were updated by adding this batch of new detectors were named with a prefix "DT3_"._ 
=> screenshot does not show the new detectors DT3_

- _Depending on how many new readers there are, it may be worthwile to develop an automated process in PostgreSQL, but this update was done manually in the Excel template. In addition to the lat/lon, the streets where the readers are is also needed in 2 columns, and the segment name._ 
=> What are the 2 columns called? There is no "segment name" in the template. 

- _The segment name is always the first two letters of each street, with the corridor street first and the intersecting street second. For example, a reader at Bloor/Christie measuring travel times on Bloor would be named `BL_CH`. For each proposed route, the excel sheet is populated with the `start reader`, `end_reader` and assigned a unique `analysis_id`. For this batch of new readers, analysis ids starting from 1600000 were added._ 
=> There are no columns `start reader`, `end_reader` and `analysis_id` in the template.

- _Therefore the reader table will have the following fields populated:  `analysis_id`, `street`, `direction`, `from_street`, `to_street`, `from_id`, `to_id`, `start_point_lat`, `start_point_lon`, `end_point_lat` and `end_point_lon`._
=> these columns are not in the screenshot. Where are they supposed to be added? Also, there are many columns in the template that are not explained. e.g. `!ozone`, `!zonelist`, `Maxtime`, etc

- _After uploading excel data to PostgreSQL_
=> to where? `bluetooth.all_analyses` ?

- Should the query `SELECT DISTINCT mohan.new_added_detectors.from_id::integer AS bluetooth_id`  have `CREATE TABLE mohan.bluetooth_nodes AS` at the beginning?

- _especially for oddly shaped intersections._
=> can you give an example of oddly-shaped intersections? What happens with them? How to fix them?

- _The table is now ready to be appended to the existing routes table._ 
=> can you show the query to append the table?

Commit:
#381 Readme file updated for creating new routes after Cathy's comments

Commit:
#381 Readme file updated for creating new routes

Commit:
#326 Readme file updated for new tables  after Raphael's comments

Commit:
#343 blip pipeline fail check slack message in new format

Commit:
#343 blip pipeline fail check syntax corrected

Commit:
#343 blip pipeline check syntax update

Commit:
#343 blip pipeline fail check syntax updated

Commit:
#343 blip pipeline connection corrected

Commit:
#343 blip pipeline start_end  corrected

Commit:
#343 blip pipeline ds corrected for yesterday

Commit:
#343 blip pipeline ds corrected for yesterday

Commit:
#343 blip pipeline ds corrected

Commit:
#343 blip pipeline ds updated

Commit:
#343 blip pipeline updated

Commit:
#343 blip pipeline syntax error corrected

Commit:
#343 blip pipeline functions schema access updated

Commit:
#343 blip pipeline functions updated

Commit:
#343 blip pipeline ds  updated with spaces

Commit:
#343 blip pipeline ds  python operator updad

Commit:
#343 blip pipeline ds  python operator updated

Commit:
#343 blip pipeline ds python operator updated

Commit:
#343 blip pipeline ds python operator updated

Commit:
#343 blip pipeline ds python operator updated

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+        select_query1 = '''SELECT MAX (datetime_bin)::date from bluetooth.aggr_5min'''
+        cursor.execute(select_query1)
+        latest_date = cursor.fetchone()
+        if (latest_date[0]) >= (datetime.now().date() - timedelta(1)):
```
This should be using `{{ ds }}` as well.

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+    op_kwargs={
+        'con': con
+        })
+    task2 = PostgresOperator(sql='''SELECT * from mohan.insert_report_date()''',
```
This should probably be using `{{ ds }}`

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+#to connect to pgadmin bot
+bt_postgres = PostgresHook("bt_bot")
+
+#To connect to pgadmin
+#CONFIG = configparser.ConfigParser()
+#CONFIG.read(r'/home/mohan/cre.cfg')
+#dbset = CONFIG['DBSETTINGS']
```
these should be deleted

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+                }
+with DAG('blip_check_update', default_args=default_args, schedule_interval='0 17 * * *') as blip_pipeline:
+    task1 = PythonOperator(
+    task_id = 'pipeline_check',
+    python_callable = pipeline_check,
+    dag=blip_pipeline,
+    op_kwargs={
+        'con': con
+        })
```
This should be tabbed over properly for readability

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+                            retries = 0,
+                            dag=blip_pipeline)                        
+    task5 = PythonOperator(
+    task_id = 'broken_readers',
+    python_callable = broken_readers,
+    dag=blip_pipeline,
+    op_kwargs={
+        'con': con,
+        'check_date': '{{ ds }}'
+        }) 
```
This should be tabbed over properly for readability

radumas commented on dags/check_brokenreaders.py:
```diff
@@ -0,0 +1,122 @@
+        'con': con,
+        'check_date': '{{ ds }}'
+        }) 
+task1 >> [task2, task3, task4, task5]
```
Are tasks 2 to 5 really independent of each other?

radumas reviewed (changes requested):
We're getting there! Close to functioning automated processes.

This section needs to be updated to discuss the updated functions you've written. 

https://github.com/CityofToronto/bdit_data-sources/blame/5dad382a6f709d9c397148008000b8768bab90ec/bluetooth/README.md#L301-L308

When the DAG works we're going to have to change everything to run from the `bluetooth` schema.

webgisgeek commented on bluetooth/README.md:
```diff
@@ -167,6 +171,79 @@ substring(cod::bit(24) from 17 for 6) as minor_device_class,
+
+#### reader_locations 
+
+This is a table of all locations at which Bliptrack readers have been installed and are physically existing. The installed readers could be online or offline but has NOT been removed physically. Each intersection has only **one** reader that is assigned to a route/routes. Therefore, if there are more than one readers in a locations that has not been removed, such detectors are listed in the `read ...
```
@radumas. There is a lookup table named reader_status_history which is also updated by the function and is included as one of the tasks in DAG. Is it necessary to include that table as well ?

radumas replied:
yep

Commit:
#343 blip pipeline slack channel updated

Commit:
#343 blip pipeline date in python operator fixed

Commit:
#343 blip pipeline date in syntax error fixed

Commit:
#343 blip pipeline syntax error fixed

Commit:
#343 blip pipeline syntax error fixed

Commit:
#343 blip pipeline quick fix for pipeline check

Commit:
#263 Bluetooth readme.md updated with lookup table addition

Commit:
#317 Bluetooth update  readme.md updated with liinks

Commit:
#343 empty list for broken reader test condition resolved

radumas commented:
`check_brokenreaders.py` should be renamed to `bluetooth_check_readers.py` so that the data source is the first word. I should update the documentation to specify a standard for how we name DAGs.

Commit:
#250 functions changed to use bluetooth schema instead of personal schema

Commit:
the detector_history_corrected table explaineaiained

webgisgeek committed:
Offline bluetooth readers past one week.

There is a view mohan.offline_readers which compares the current reader status (signal last received in within one week as 'online' else 'offline') vs the status of bt readers in the reader_history table which we received from JJ ('past status')

Commit:
#381 DAG name corrected

Commit:
#343 DAG name corrected and PR feedback incorporated

Commit:
Merge branch 'btdag' of https://github.com/CityofToronto/bdit_data-sources into btdag

radumas commented on bluetooth/README.md:
````diff
@@ -167,6 +172,91 @@ substring(cod::bit(24) from 17 for 6) as minor_device_class,
 substring(cod::bit(24) from 12 for 5) as major_device_class
 ```
 
+As a part of updating/modernizing Bluetooth Reader Tables [ISSUE #196](https://github.com/CityofToronto/bdit_data-sources/issues/196), the following four tables are prepared.  
+* `reader_history`
+* `reader_locations`
+* `routes`
+* `reader_status_history`
````
I know this is supposed to be introducing the new tables you've created, but this comment doesn't make sense where it is, which is under the `ClassOfDevice` heading.
This should be placed under the `Internal Tables`, and maybe move the sections for the individual tables above the `ClassOfDevice` section

radumas reviewed (changes requested):
We're getting there.

There should only be 1 DAG in this repository, however, please delete the other one. The proper way to rename a file is to use `git mv` to reduce the effect on the history.

There needs to be an explanation of what that DAG does under section 5 of the main README: explaining what the purpose of the validation routines are.

Please carefully review the comments on what's been committed, I've noticed some still haven't been acted upon. Let me know if you have questions about them.

webgisgeek committed:
Delete bt_read_history.py

a trial file not needed anymore.

webgisgeek committed:
Delete bt_test_dag.py

test file deleted

webgisgeek committed:
Delete check_brokenreaders.py

Not relevant after bluetooth_check_readers.py

radumas commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,118 @@
+			with x AS 
+			(select DISTINCT(analysis_id), 
+			 CASE
+			 WHEN MAX (datetime_bin::date)>= (insert_value-1) then (insert_value-1)
```
This comparison is wrong. `insert_value` is the `execution_date`, which would be the day over which the datetime_bins came in. So there shouldn't be `-1` in this comparison.

radumas commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,37 @@
+where bluetooth.reader_status_history.active IS false and (bluetooth.reader_status_history.dt = insert_dt and last_active_date = insert_dt-2)
+
+)
+INSERT INTO bluetooth.broken_readers_log(reader_id, reader_name, check_date)
+select reader_id, name, dt
+from y
+left join bluetooth.reader_locations using(reader_id)
```
This `broken_readers_log` is redundant. The same information can be gleaned from the `reader_status_history` table.
Also, this function doesn't return anything, it is only `INSERT`ing, so the check in the DAG task won't do anything.

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+                'on_failure_callback': task_fail_slack_alert
+                }
+with DAG('bluetooth_check_readers', default_args=default_args, schedule_interval='0 17 * * *', catchup=False) as blip_pipeline:
+    task1 = PythonOperator(
```
Rename the `taskN` Python variables with their `task_id` so it's easier to understand what each one is doing.

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+        'con': con,
+        'check_date': '{{ ds }}'
+        }) 
+task1, task2, task3, task4, task5
```
The `broken_readers` task depends on the `reader_status_history` one, they can't happen independently.

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+        select_query2 = '''SELECT * from bluetooth.broken_readers(%s::date)'''
+        cursor.execute(select_query2, (check_date,))
+        broken_readers = cursor.fetchall()
+        broken_list.append(broken_readers)
```
This step seems unnecessary

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+        num_broken = len(broken_list)
+        #please suggest if this try except works
+        try:
+            if broken_readers[0][0] == '':
```
why not use `num_broken`?

The Slack message here should print out the broken readers

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+def task_fail_slack_alert(context):
+    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
+    
+    if context.get('task_instance').task_id == 'task1':
```
These values for `task_id` are wrong, in this first case it would `pipeline_check`

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+            <@U01858E603T> fix it asap.""".format(
+                task=context.get('task_instance').task_id,)
+    
+    elif context.get('task_instance').task_id == 'task5':
+        task_msg = """:among_us_ghost: some readers {task} have failed. Check the log, <@U01858E603T>.""".format(
+                task=context.get('task_instance').task_id,)
```
This can be deleted.

radumas commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,127 @@
+    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
+    
+    if context.get('task_instance').task_id == 'task1':
+        task_msg = """:among_us_dead: Task {task} in blip check update failed, 
```
This should read "No bluetooth data was found in the database for this date. Remote desktop to the terminal server and check the `blip_api` log".

radumas commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,118 @@
+	active.dt
+   FROM active)
+   
+   INSERT INTO bluetooth.reader_status_history (reader_id, last_active_date, active, dt)
+   SELECT DISTINCT reader_id, max(max),route_status::bool, dt
+   from final
+   where reader_id IS NOT NULL
+   group by reader_id, route_status, dt
+   ORDER BY reader_id
```
Most of the CTEs in this function and `reader_locations_dt_update` is the same. This `INSERT` can be in a CTE and then the `UPDATE` from `reader_locations_dt_update.sql` could be run below it.

Then there could be one task to do both in the DAG

radumas reviewed (changes requested):
Some changes to simplify automated tasks and some bug fixes.

webgisgeek committed:
Updated broken_readers.sql

Now this sql returns id and reader_name for broken readers from reader_status_history table.

webgisgeek committed:
Update reader_status_history.sql

Both reader_status_history and reader_locations tables updated by a single function.

webgisgeek committed:
Update README.md

Updated after PR comments.

webgisgeek committed:
Update bluetooth_check_readers.py

PR comments incorporated. Has not been tested in airflow in real time.

chmnata committed:
#450 Update slack alerts to use variable

chmnata commented:
Seems like the below two DAGs have been deleted and are no longer in the branch. I will rm them in the dag folder on the ec2 as well. 
bt_read_history.py -> /etc/airflow/dev_scripts/dags/bt_read_history.py
bt_test_dag.py -> /etc/airflow/dev_scripts/dags/bt_test_dag.py

radumas committed:
#343 bugfix

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+			z.dt
+           FROM z
+        )
+		, active AS (
```
can be included in the `final` cte

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+           FROM c
+             LEFT JOIN bluetooth.detectors_history_final ON c.detector_name = bluetooth.detectors_history_final.read_name::text
+          WHERE c.route_status = 'True'::text
+          GROUP BY c.route_status, c.detector_name, bluetooth.detectors_history_final.reader_id, c.dt
```
no need to add schema name

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+			 END 
+			 as last_reported,
+			CASE
+			WHEN  MAX (datetime_bin::date)>= (insert_value) then 'True'::text
```
this should be kept as boolean because at the very end it got casted back to boolean

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+            bluetooth.detectors_history_final.reader_id as id,
+			c.dt
+           FROM c
+             LEFT JOIN bluetooth.detectors_history_final ON c.detector_name = bluetooth.detectors_history_final.read_name::text
```
no need to cast read_name to text

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+        )
+		, active AS (
+         SELECT DISTINCT c.detector_name,
+            max(c.last_reported) AS max,
```
should be called last_reported

chmnata committed:
#381, #250 address comment for reader_status_history sql

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,122 @@
+    bluetooth.detectors_history_final.reader_id,
+			c.dt
+   FROM c
+     LEFT JOIN bluetooth.detectors_history_final ON c.detector_name = bluetooth.detectors_history_final.read_name::text
```
not exactly sure what is detectors_history_final as it is not bluetooth schema and not documented

chmnata replied:
Found 5 `detectors_history_*` tables in schema `mohan`, 
`detectors_history_final`
`detectors_history`
`detectors_history_late`
`detectors_history_latest`
`detectors_history_corrected`

chmnata replied:
seems like the reason to join to this table is to get the corresponding `reader_id` which is needed in the output. This should also be an inner join

chmnata commented on bluetooth/sql/functions/reader_locations_dt_update.sql:
```diff
@@ -0,0 +1,113 @@
+
+-- DROP FUNCTION bluetooth.reader_locations_dt_update(date);
+
+CREATE OR REPLACE FUNCTION bluetooth.reader_locations_dt_update(
```
Since this is included in reader_status_history I think this can be deleted

chmnata commented on bluetooth/README.md:
```diff
@@ -134,6 +161,72 @@ The script pulls the route configurations nightly from the Blip server. These ar
+
+Except the `reader_history_id` all other fields in this table has to be updated manually.
+
+#### reader_locations 
```
detector_history table is missing in this doc

chmnata commented:
For function `reader_status_history`, there is a join to a `detectors_history_final` table which was not documented in the readme. There are also 5 versions of it in the schema currently. The purpose of the join is to get the reader_id from `detectors_history_final` using `read_name` = `detector_name`. However, there are duplicated `read_name` in `detectors_history_final` as it contains all the past read_name for every reader_id. I don't think the historical read_name matters for this function as we are only running the route configurations that are currently in place. So I checked out another table called `reader_locations` and it seems to have the same read_name and detector_name relationship as the  `detectors_history_final` table, but without the historic values. I think this would be a more suitable table to get the reader_id unless there is a look up table somewhere else...

chmnata committed:
#381, #250 update column name

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+AS $BODY$	
+	begin
+    return query 
+    with x as(
```
cte name should be called something descriptive

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+end; $BODY$;
+
+ALTER FUNCTION bluetooth.broken_readers(date)
+    OWNER TO bluetooth;
```
bluetooth is not a group or a role, should be `bt_admins`

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+    OWNER TO bluetooth;
+
+GRANT EXECUTE ON FUNCTION bluetooth.broken_readers(date) TO PUBLIC;
+
```
Add a comment to describe what this function is doing

chmnata commented on bluetooth/sql/functions/reader_locations_dt_update.sql:
```diff
@@ -0,0 +1,113 @@
+end; $BODY$;
+
+ALTER FUNCTION bluetooth.reader_locations_dt_update(date)
+    OWNER TO bluetooth;
```
change to bt_admins

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+,y as (
+select reader_id, last_active_date as today, active as active_today
+	from bluetooth.reader_status_history
+where last_active_date = insert_dt::date - 1 and active is false
```
the active is false is not needed as it is also in the where clause in the below cte

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+where last_active_date = insert_dt::date - 1 and active is false
+order by reader_id
+)
+select distinct (reader_id), read_name from x 
```
I think this can be simplified to
WHERE active is False and last_active_date was yesterday. Aka the reader is currently not active but the last active day was yesterday. Then we will not need any ctes from above.

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+)
+select distinct (reader_id), read_name from x 
+left join y using(reader_id)
+left join bluetooth.detectors_history_corrected using(reader_id)
```
left join also results in duplicated rows, with one reader_id and multiple reader_names even when some are offline

chmnata commented on bluetooth/sql/functions/broken_readers.sql:
```diff
@@ -0,0 +1,40 @@
+
+GRANT EXECUTE ON FUNCTION bluetooth.broken_readers(date) TO PUBLIC;
+
+GRANT EXECUTE ON FUNCTION bluetooth.broken_readers(date) TO bluetooth;
```
this should be either bt_bot or bt_admin_bot

chmnata replied:
going to put it as bt_bot as of now

chmnata committed:
#385, #343 Simplify broken_readers.sql and add comments

chmnata committed:
#385 Add missing insert_report_date function

chmnata commented on bluetooth/sql/functions/insert_report_date.sql:
```diff
@@ -0,0 +1,28 @@
+AS $BODY$
+	
+	begin
+		with x AS 
```
change to descriptive cte name

chmnata committed:
#343 Update insert report date sql used in the dag

chmnata commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,119 @@
+broken_list = []
+# check if the blip data was aggregated into aggr_5min bins as of yesterday.
+# if the aggr_5min table has at least one record for yesterday, it means pipeline is working 
+def pipeline_check(con, check_date):
```
!!! I can we can use SQLcheckoperator here !!!!

chmnata committed:
#343 Update pipeline check to use sqlcheckoperator and add comments

chmnata committed:
#385 update broken readers sql

chmnata committed:
#381, #343 Update broken reader task to alert and not kill self

chmnata commented on bluetooth/sql/create_tables/add_new_segments.sql:
```diff
@@ -0,0 +1,31 @@
+FROM ( 
+SELECT analysis_id, street, direction, from_street, to_street, 
+		gis.twochar_direction(gis.direction_from_line(ST_linemerge(ST_union(geom)))) AS geom_dir,
+		ST_linemerge(ST_union(geom)) AS geom
```
the table should also have a length column

chmnata committed:
#381, #317 update segment table to include length column

chmnata committed:
#131 Update blip api to use psycopg2 instead of pg

chmnata committed:
#131 Updpg json encoder to use json dump

chmnata committed:
#131update comments from pg to psycopg2

chmnata committed:
#131 update upsert sql

chmnata committed:
#131fix that json mess in prepping upsert row

chmnata committed:
#131 fix upsert sql for column name and got rid of included column

chmnata committed:
#131 change from tuple to index for formatting analyses_pull_data

chmnata committed:
#131 got rid of print() comments

chmnata committed:
#131 fix upsert returned values to use inserting row

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,117 @@
+$BODY$;
+
+ALTER FUNCTION bluetooth.reader_status_history(date)
+    OWNER TO bluetooth;
```
owner to bt_admins

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
+`to_id`				- The four digit Unique bluetooth id at the end point
+`to_lat`			- Latitude at the end point of the route
+`to_lon`			- Longitude at the end point of the route
 
```
add a length column

chmnata commented on bluetooth/update/README.md:
````diff
@@ -2,145 +2,133 @@
-```
 
-Matches the corridors to the names based on similar from intersections.
+This .csv file was then imported into QGIS and pushed into the mohan schema in postgres. The geometry column is created by adding a column using `ALTER TABLE mohan.new_added_detectors ADD COLUMN from_geom GEOMETRY` we need two geometry columns: `from_geom` and `to_geom`.  Filling in the from_geom column with `UPDATE mohan.new_added_detectors SET to_geom = ST_SetSRID(ST_MakePoint(to_lon, to_lat),  ...
````
I feel like the csv can be imported directly to postgres instead of going through QGIS, since QGIS function wasn't used

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
 
-Makes a single table for the corridors and readers at the from/to intersections.
+To get the intersection ids that are close to the newly added detectors location, create a table named `bluetooth_nodes`. This table has four fields:
+`bluetooth_id`, `geom` (this is geometry of bluetooth detectors), `int_id` (nearest intersection id) and `int_geom` (geometry of the nearest intersection to the )
```
unfinished sentence

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
+     CROSS JOIN LATERAL ( SELECT z.int_id,
+            st_transform(z.geom, 98012) AS node_geom
+           FROM gis.centreline_intersection z
+          ORDER BY (z.geom <-> mohan.new_added_detectors.from_geom)
```
would we miss detectors because only from detector's geometry was used?

chmnata replied:
feel like we should be using a distinct detector id of both from and to detector

chmnata commented on bluetooth/update/README.md:
````diff
@@ -2,145 +2,133 @@
+         LIMIT 1) nodes);
 ```
-Draws lines using the reader locations. This is the result.
+Check that correct intersections are returned from this query especially for oblique intersections with an offset. If required, correct the intersection_id and geom for such intersections and finalize the table `mohan.bluetooth_nodes`. An example is shown below. The intersection between Dundas St W, Roncesvalles Ave and Boustead Ave is oblique with an offset. The BT reader 5263 (red dot) picked u ...
````
extra 1 in `113466258`

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
-INTO rliu.new_bt_coflate1000 
-FROM conflation
+SELECT analysis_id, street, direction, from_street, to_street,
+	CASE WHEN geom_dir != direction THEN ST_reverse(geom) 
```
not sure if this is needed because we are using centreline with direction as a base network for which the direction of the line is already flipped. This might result in double flipped geometries if the dirs didnt match because of a NB vs WB comparison

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
 
-![bt_segments](img/bt_segments.PNG)
+## Things to note 
+Geostatistical lines and planning boundaries need to be avoided while pgrouting. 
```
also railways and rivers and stuff

chmnata commented on bluetooth/update/README.md:
```diff
@@ -2,145 +2,133 @@
 
-## Cleanup, Warnings, and Alternative Methods
+## Validating Output
+Validate the length of the segments with length ST_length(geom) and direction using gis.direction_from_line(geom) functions.If the detectors are located very close to the centerline intersections, it is not necessary to do the  centreline cutting. If any bluetooth detectors are not located at the start or end point of a centreline, we will need to cut the centreline using ST_linelocatepoint() as  ...
```
I think it would be helpful to include the steps to cut the line here in the readme instead of linking the issue.

chmnata committed:
#381 edit update readme

chmnata committed:
#381 change owner to bt_admins

chmnata committed:
#381 add st_linesubstring steps to readme

chmnata commented on bluetooth/update/README.md:
````diff
@@ -2,145 +2,173 @@
+```
 
-At this point, the layer has multiple entries containing centreline segments for each analysis_id. Instead, we need a single line/entry for each analysis_id. This query concatenates the individual centreline segments into one line. The main function doing this work is [`ST_LineMerge()`](https://postgis.net/docs/ST_LineMerge.html)
+The new routes table is now ready to append to the existing routes table.  
````
The `bluetooth.segments` table has the columns:     

- segment_name ,
- analysis_id,
- street ,
- direction,
- start_cross ,
- end_cross ,
- length ,
- bluetooth ,
- wifi ,
- geom ,
- duplicate ,
- reversed ,
- end_street

So its missing: 
- segment_name , bluetooth, widi, duplicate, revered, end_street

chmnata commented on bluetooth/update/reader_status_details.md:
```diff
@@ -0,0 +1,40 @@
+# Bluetooth reader's status
```
Maybe this can go under the readme?

chmnata commented:
List of modified things:
SQL
- Functions:
	- broken readers: simplified, need to know which version of `detectors_history_corrected` should be used
	- insert report date: simplified 
	- reader locations dt update: should be deleted
	- reader status history: simplified, need to know which version of `detectors_history_final` should be used
- Table:
	- added length column to mohan.bt_segments_new
	- OWNER to bt_admins 
	- Change create new segment to not ST_reverse geom

DAG: 
	- updated to use sqlsensor
	- updated to send alert message for broken readers and not fail the task

README
	- added sql examples in readme
	- added steps to cut lines with linesubstring

Things that needs to be fixed:
- readme is missing the explanation of detectors_history, currently in another md under update
- mohan.bt_segments_new is not ready to insert to bluetooth.segments, its missing a couple columns
- need to figure out which version of `detectors_history_corrected` to use then update function: broken reader and reader status history 
- sqls are only updated in github not in postgres
-

chmnata committed:
#381 update create new segment to not use ST_reverse for geom

chmnata committed:
#381 update reader status history to use correct version of table

chmnata committed:
#381 move updating readme to main readme

chmnata commented:
Looking to move the `detectors_history_*` to bluetooth schema, `mohan.detectors_history_corrected` is the corrected version, however there are some null values on several columns.  I will dig into how to populate those. 
e.g. some geoms are null and bt_id are null 
![image](https://user-images.githubusercontent.com/46324452/162042442-593c79d6-92f0-4fc2-9e0b-eb59c722c8dc.png)

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,117 @@
+      
+      -- Combine active and deactive detector 
+      , detector_status as (
+         SELECT      DISTINCT detector_list.detector_name,
```
should select distinct on reader_id instead of detector_name, because we are inserting reader_id's status at the end of the query and multiple detector_names can be refer to one reader_id
```
-- return 216 rows
select distinct reader_id, read_name from mohan.detectors_history_corrected

-- return 178 rows
select distinct reader_id from mohan.detectors_history_corrected

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,117 @@
+   
+   INSERT INTO       bluetooth.reader_status_history (reader_id, last_active_date, active, dt)
+
+   SELECT            DISTINCT reader_id, max(last_reported), route_status, dt
```
I don't think this select statement is necessary because the previous cte already remove possible dups

chmnata commented on bluetooth/sql/functions/reader_status_history.sql:
```diff
@@ -0,0 +1,117 @@
+
+   SELECT            DISTINCT reader_id, max(last_reported), route_status, dt
+   FROM              detector_status
+   WHERE             reader_id IS NOT NULL
```
there is no null reader_id in the detector_history table

chmnata committed:
#381 update reader_status_history function

chmnata commented:
Working on the new segments tables and adding the missing columns: segment_name , bluetooth, wifi, duplicate, reverse, end_street
For bluetooth and wifi, according to https://github.com/CityofToronto/bdit_data-sources/blob/e98182d30de5511a2a4c246e4771f14c937c6b0f/bluetooth/sql/create_tables/segments.sql seems like bluetooth was assigned TRUE and wifi was assigned FALSE, but when I check the bluetooth.segments table there are values for wifi that are true 😕 
Also not sure what reversed means, reversed geometry (?)

chmnata committed:
#381 update bt new segments table with new columns

radumas commented:
Just in case future spelunkers come here.
`wifi = TRUE` on highways where the problems associated with using WIFI observations are reduced.
`reversed` was a flag to identify segments had been drawn against the direction of travel and which had been subsequently reversed. I'm really not sure why it stuck around.
`duplicate` appears to have been to identify overlapping segments, which primarily exist on College. Since A LOT of those routes are offline, and some appear to have never been created, I am rather confused by the state of those detectors.

chmnata committed:
#343 update check to not use case when

chmnata commented:
In bluetooth.routes_temp, there are 61 routes out of 402 with no data in `bluetooth.aggr_5min` and `bluetooth.observations`. 

![image](https://user-images.githubusercontent.com/46324452/165591044-b1718559-3b24-472b-aa16-dfce77372ea5.png)


But it does seems like some routes with data do have coverage on those routes with no data. (Green dotted lines are the ones without data and grey lines are the ones with data) 
![image](https://user-images.githubusercontent.com/46324452/165594625-b1649467-fe7f-4ad2-8f24-d253a78440ce.png)
![image](https://user-images.githubusercontent.com/46324452/165595304-30b95835-3e6a-406a-af9a-218dd03f0649.png)

Ones that still have coverage (or partial coverage):
- Adelaide from parliament to broadview 1495398, covered by 1454095
- Parliament from gerrard to calton 1496472, covered by 1455692 (Parliament from dundas to calton)
- College from ossington todufferin (1600090), sorta half covered by college from dovercourt to bathurst 1412888
- Yonge between Chatsworth and Castlefield (1460798, 1461616), covered by various short and long routes (1428295, 1448887, 1448891, 1448413, 1428324, 1448861)
- Sheppard from Morningside Ave to Nelson (end street was coded as sheppard incorrectly) 1460834, covered by VARIOUSSSSS short and long routes 5 in total

chmnata commented:
Checking the distinct status in reader_history table, `routepoint name` does not sounds like a status type (?), `added` is also ???. I guess technically according to this issue https://github.com/CityofToronto/bdit_data-sources/issues/196, this table doesn't need a status column 🤔 
- "duplicate"
- "routepoint name"
- "online"
- "offline"
- "decommissioned"
- "added"
- "moved"

The following readme explained the meaning behind each status, kindddd of feels like the below 3 is confusing or just keeping a log of duplicated data/data of human error. 
https://github.com/CityofToronto/bdit_data-sources/blob/btdag/bluetooth/update/reader_status_details.md

- added (these are the locations where names are inconsistent for example a reader with bt_id 4035 has a name E in one table and E4035 in another. Despite being at the same intersection at Gardiner & Parliament. Therefore, these two names could refer to the same reader. Thus to distinguish one got the status added and another is offline )

- duplicate (duplicate does not have a bt_id but based on its location and/or the reader name, there are more than one at a given intersection. The one that has got less details is assigned as duplicate)

- moved (moved has a bt_id but the same bt_id existed in a different location in the past. For example bt_id 5331 was at Gerrard & University. But the same bt_id also is at Lawrence & Mt Pleasant and is offline as per latest table. Thus 5331 has two entries with status offline and moved.

- routepoint name (does not have a bt_id. It is a name given to a point along a route. Could there be a detector at those points ? most likely not)

chmnata committed:
#343 Update bt dag pipeline check task to be idempotent

chmnata committed:
#250 update function

chmnata commented on dags/bluetooth_check_readers.py:
```diff
@@ -0,0 +1,128 @@
+from attr import NOTHING
```
outdated, remove

chmnata commented on bluetooth/api/blip_api.py:
```diff
@@ -12,16 +9,20 @@
 import logging
 import os
 import sys
+#from syslog import LOG_DEBUG
```
updated in more recent PRs. outdated, remove
https://github.com/CityofToronto/bdit_data-sources/pull/642
https://github.com/CityofToronto/bdit_data-sources/pull/480

chmnata commented on bluetooth/sql/create_tables/add_new_segments.sql:
```diff
@@ -0,0 +1,80 @@
+CREATE table mohan.bt_segments_new AS 
```
The segments in there that does not exist in `bluetooth.routes_temp` (when joining with `analysis_id`) are the following: 

```
'Bloor'	'WB'	'HighPark'	'Runnymede' 
'Bathurst'	'SB'	'Bloor'	'Harbord'
'Bloor'	'EB'	'Runnymede'	'HighPark'
'Queens'	'NB'	'Wellesley'	'Bloor
'Queens'	'SB'	'Bloor'	'Hoskin
'Spadina'	'NB'	'College'	'Harbord'
'Bathurst'	'NB'	'Harbord'	'Bloor'
'Spadina'	'SB'	'Harbord'	'College'
```
They are actually in `bluetooth.routes_temp`, just under a different analysis_id. I assume the one in `bluetooth.routes_temp` is correct because there is data in `bluetooth.aggr_temp` for those segments, and when searching for data with the analysis_id in mohan's table, no data returned.

```
-- returned none
SELECT * FROM bluetooth.aggr_5min where analysis_id in (1600060,
1600030,
1600070,
1600250,
1600220,
1600290,
1600010,
1600300)
```

i believe it is safe to delete this file

chmnata commented on bluetooth/readersdown/offline_readers.csv:
```diff
@@ -0,0 +1,22 @@
+reader_id,bt_id,past_status,current_status,name,int_id,project_name,geom
```
@radumas do we even need this in the repo? should it be in a K drive or sth instead?

radumas replied:
Aye could be in K `Data Sources\Bluetooth\Bliptrack\Inventory` and include the date it was committed in the filename
